File: python/allergies/allergies.py
class Allergies:
    allergies = [
        "eggs",
        "peanuts",
        "shellfish",
        "strawberries",
        "tomatoes",
        "chocolate",
        "pollen",
        "cats",
    ]

    def __init__(self, score):
        self.score = score

    def allergic_to(self, item):
        return self.score & (1 << self.allergies.index(item)) > 0

    @property
    def lst(self):
        return [allergy for allergy in self.allergies if self.allergic_to(allergy)]

File: haskell/allergies/src/Allergies.hs
module Allergies
  ( Allergen(..)
  , allergies
  , isAllergicTo
  ) where

import           Data.Bits                      ( testBit )

data Allergen = Eggs
              | Peanuts
              | Shellfish
              | Strawberries
              | Tomatoes
              | Chocolate
              | Pollen
              | Cats
              deriving (Bounded, Enum, Eq, Show)

allergies :: Int -> [Allergen]
allergies score = filter (`isAllergicTo` score) [minBound ..]

isAllergicTo :: Allergen -> Int -> Bool
isAllergicTo allergen score = testBit score (fromEnum allergen)
